cedar-policy/src/api/err.rs

```diff
@@ -978,3 +978,189 @@ impl From<cedar_policy_core::ast::RestrictedExpressionParseError>
+    UndeclaredPrincipalType(#[from] request_validation_errors::UndeclaredPrincipalTypeError),
+    /// Request resource is of a type not declared in the schema
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    UndeclaredResourceType(#[from] request_validation_errors::UndeclaredResourceTypeError),
+    /// Request principal is of a type that is declared in the schema, but is
+    /// not valid for the request action
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    InvalidPrincipalType(#[from] request_validation_errors::InvalidPrincipalTypeError),
+    /// Request resource is of a type that is declared in the schema, but is
+    /// not valid for the request action
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    InvalidResourceType(#[from] request_validation_errors::InvalidResourceTypeError),
+    /// Context does not comply with the shape specified for the request action
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    InvalidContext(#[from] request_validation_errors::InvalidContextError),
+    /// Error computing the type of the `Context`
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    TypeOfContext(#[from] request_validation_errors::TypeOfContextError),
+}
+
+#[doc(hidden)]
+impl From<cedar_policy_validator::RequestValidationError> for RequestValidationError {
+    fn from(e: cedar_policy_validator::RequestValidationError) -> Self {
+        match e {
+            cedar_policy_validator::RequestValidationError::UndeclaredAction(e) => {
+                Self::UndeclaredAction(e.into())
+            }
+            cedar_policy_validator::RequestValidationError::UndeclaredPrincipalType(e) => {
+                Self::UndeclaredPrincipalType(e.into())
+            }
+            cedar_policy_validator::RequestValidationError::UndeclaredResourceType(e) => {
+                Self::UndeclaredResourceType(e.into())
+            }
+            cedar_policy_validator::RequestValidationError::InvalidPrincipalType(e) => {
+                Self::InvalidPrincipalType(e.into())
+            }
+            cedar_policy_validator::RequestValidationError::InvalidResourceType(e) => {
+                Self::InvalidResourceType(e.into())
+            }
+            cedar_policy_validator::RequestValidationError::InvalidContext(e) => {
+                Self::InvalidContext(e.into())
+            }
+            cedar_policy_validator::RequestValidationError::TypeOfContext(e) => {
+                Self::TypeOfContext(e.into())
+            }
+        }
+    }
+}
+
+/// Error subtypes for [`RequestValidationError`]
+pub mod request_validation_errors {
+    use miette::Diagnostic;
+    use ref_cast::RefCast;
+    use thiserror::Error;
+
+    use crate::{Context, EntityTypeName, EntityUid};
+
+    /// Request action is not declared in the schema
+    #[derive(Debug, Diagnostic, Error)]
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    pub struct UndeclaredActionError(
```

Comment:
If only there were also `#[doc(transparent)]`.

Reply:
Sadly it [doesn't look like that exists](https://doc.rust-lang.org/rustdoc/write-documentation/the-doc-attribute.html)